# Pre-Lecture HW
# 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [2]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

# 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then

### 1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,

In [3]:
rows, columns = df.shape

print(f"The dataset has {rows} rows and {columns} columns.")

The dataset has 391 rows and 11 columns.


### 2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

### Answer:
Observations are a bunch of object or individual data that represent the rest of data in that row. For my data set, the observations are Admiral, Agent S, Agnes, etc. While variables are the attributes or things that describe the observations. Each column represents different variables. For my dataset, the variables are id, gender, species, etc.

# 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [4]:
# Summary of numerical columns
numerical_summary = df.describe()

# Summary of all columns (gives info on data types, non-null counts, etc.)
general_info = df.info()

# Summary of categorical columns
categorical_summary = df.select_dtypes(include='object').apply(pd.Series.value_counts)

# Print the summaries
print("Numerical Summary:\n", numerical_summary)
print("\nGeneral Info:\n", general_info)
print("\nCategorical Summary:\n", categorical_summary)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
Numerical Summary:
             row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

General Info:
 None

Categorical Summary:
             id  name  gender  species  birthday  personality  song  phrase  \
1-1   

# 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
tc = pd.read_csv(url)
tc

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [6]:
tc.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [7]:
tc.shape

(891, 15)

In [8]:
tc.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

### Answer:
There are discrepancies between the size of the dataset reported by tc.shape and tc.describe(). These discrepancies are:

1) There are 15 columns reported by tc.shape, while tc.describe() shows only 6 columns. This happens because tc.shape counts all of the columns in the dataset, regardless of the data type of each column. Meanwhile, tc.describe() only counts and shows columns that have numeric data types.

2) There are 891 rows reported by tc.shape. In the "count" row of tc.describe(), all columns have a value of 891 except for the age column, which has a value of 714. This occurs because tc.describe() doesn't count cells with missing values in the age column. This explains why the value in the count row for the age column is 714. This number is derived from the total number of rows in the dataset (891) minus the number of missing values in the age column (177), resulting in 891 - 177 = 714.

# 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

1.  an "attribute", such as df.shape which does not end with ()
and 
2. "method", such as df.describe() which does end with ()

### Answer: 
An attribute is a characteristic or piece of data associated with a dataset, while a method is a function that is associated with a dataset and can perform operations or computations using data from the dataset.

# Summary
1. **Loading Data and Checking Shape**:
   - You asked for Python code to determine the number of rows and columns in a dataset using Pandas. I provided code that loads the dataset and checks its shape using `df.shape`.

2. **Observations vs. Variables**:
   - I explained that in your dataset, **observations** are individual rows (each representing a villager), and **variables** are columns (representing attributes like name, species, etc.).

3. **Generating Column Summaries**:
   - You wanted to summarize the columns in your dataset. I provided code to get both numerical and categorical summaries using `.describe()`, `.info()`, and `.value_counts()` methods.

4. **Attribute vs. Method**:
   - I explained that an **attribute** (like `df.shape`) gives you stored information about an object, while a **method** (like `df.describe()`) performs some action or computation.
   - I further simplified the explanation using a toy box analogy to clarify the difference between attributes and methods in a way a 7-year-old could understand.
   
### ChatGPT's link: https://chatgpt.com/share/42ae47e9-b8e6-4704-bd07-5dadb41671ce 

# Post-Lecture HW
# 6. The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

### Answer:
- 'count' gives us the total number of data points in a dataset.



- 'mean' is the average of the dataset. The value can be obtained by adding the values of all data points and then dividing by the total number of data points.


- 'std' or standard deviation is the amount of variation or spread in the data. A low standard deviation value means that the data in the dataset tend to be close to the mean, while a high standard deviation value means that the data in the dataset are spread out over a wider range.


- 'min' or minimum gives us the lowest value in the dataset.

- '25%' or the first quartile is the lowest value where the lowest 25% data in the dataset is below this value.

- '50%' or median is the middle value of the dataset when we arrange the data in an ascending order.

- '75%' or third quartile is the lowest value where the lowest 75% data in the dataset is below the value.

- 'max' or maximum gives us the highest value in the dataset.

# 7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

### 1. Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']

### Answer: 
Imagine we’re working with a dataset of customer information for an e-commerce platform. The DataFrame (df) includes columns like 'CustomerID', 'Name', 'Email', and 'PhoneNumber'.

We want to clean the dataset by removing rows where important contact information (like the email or phone number) is missing. 

Then we should df.dropna() rathen than del df['col'] because we want to delete a row that have missing values like Email or PhoneNumber. If we use del df['col'], then we will delete all of the column that we choose, for example in our case if we type del df['Email'] , all of cells in 'Email' column, whether it has value or not, will be deleted and we don't want that happened.

In [15]:
import pandas as pd

data = {'ID': [1, 2, 3, 4],
        'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Email': ['alice@example.com', None, 'charlie@example.com', None],
        'Phone': ['123-4567', '987-6543', None, '555-1212']}

ab = pd.DataFrame(data)
ab

ID     Name                Email     Phone
0   1    Alice    alice@example.com  123-4567
1   2      Bob                 None  987-6543
2   3  Charlie  charlie@example.com      None
3   4    David                 None  555-1212

In [17]:
ab.dropna()

ID   Name              Email     Phone
0   1  Alice  alice@example.com  123-4567

In [18]:
del ab['Email']
ab

ID     Name     Phone
0   1    Alice  123-4567
1   2      Bob  987-6543
2   3  Charlie      None
3   4    David  555-1212

### 2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()

### Answer:
Imagine we're working with a large dataset containing employee information. The DataFrame (df) includes columns like 'EmployeeID', 'Name', 'Department', and 'SocialSecurityNumber'.


We don't need the 'SocialSecurityNumber' column for our analysis because it's private information, and it's irrelevant to our task of analyzing department performance. So, we want to remove that column entirely from our DataFrame.


Using del df['SocialSecurityNumber'] is the better option compared to df.dropna() because our target is to delete the entire 'SocialSecurityNumber' column, not delete rows that contain empty values like df.dropna() does.

In [49]:
import pandas as pd

data = {'EmployeeID': [1001, 1002, 1003, 1004],
        'Name': ['John', 'Sara', 'Mike', 'Anna'],
        'Department': ['HR', 'Finance', 'IT', 'Marketing'],
        'SocialSecurityNumber': ['123-45-6789', None, '987-65-4321', '555-55-5555']}

cd = pd.DataFrame(data)
cd

EmployeeID  Name Department SocialSecurityNumber
0        1001  John         HR          123-45-6789
1        1002  Sara    Finance                 None
2        1003  Mike         IT          987-65-4321
3        1004  Anna  Marketing          555-55-5555

In [47]:
del cd['SocialSecurityNumber']
cd

EmployeeID  Name Department
0        1001  John         HR
1        1002  Sara    Finance
2        1003  Mike         IT
3        1004  Anna  Marketing

### 3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important

### Answer:
Usually, when we use del df['col'] in a dataset, the column we want to delete is not important. When we use df.dropna(), we want to delete rows that contain missing values. If we use df.dropna() first, rows that contain missing values in the column we want to delete will be removed as well. However, we may still want to keep these rows because they don't actually have missing values in the important columns; the missing values are in the column we intend to delete anyway. That's why we should use del df['col'] first and then use df.dropna() . 

### 4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

### Answer:
Imagine we want to delete 'Phone' column and missing data in the other column except the 'Phone' column. Then the right combination is use the del list['Phone'] before list.dropna()

In [50]:
import pandas as pd

data = {'ID': [1, 2, 3, 4],
        'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Email': ['alice@example.com', None, 'charlie@example.com', None],
        'Phone': ['123-4567', '987-6543', None, '555-1212']}

list = pd.DataFrame(data)
list

ID     Name                Email     Phone
0   1    Alice    alice@example.com  123-4567
1   2      Bob                 None  987-6543
2   3  Charlie  charlie@example.com      None
3   4    David                 None  555-1212

#### Below is the result if we use del list['Phone'] before list.dropna()

In [53]:
list = pd.DataFrame(data)
del list['Phone']
list.dropna()

ID     Name                Email
0   1    Alice    alice@example.com
2   3  Charlie  charlie@example.com

#### Below is the result if we use list.dropna() before del list['Phone']

In [57]:
list = pd.DataFrame(data)
list.dropna(inplace=True)
del list['Phone']
list

ID   Name              Email
0   1  Alice  alice@example.com

### ChatGPT's link for question 6 and 7 : https://chatgpt.com/share/66e349ec-3f84-800a-926b-91934fd6ca33
### The summary is in the ChatGPT

# 8. Give brief explanations in your own words for any requested answers to the questions below

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
tc = pd.read_csv(url)
tc

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

### Answer:

In [6]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
tc = pd.read_csv(url)
tc.groupby("pclass")["fare"].describe()

count       mean        std  min       25%      50%   75%       max
pclass                                                                     
1       216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
2       184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
3       491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

The DataFrame above is the output of tc.groupby("pclass")["fare"].describe(). First, we take the "pclass" column and use groupby to group the same values in the "pclass" column. We get "1", "2", and "3" as the different values that exist in the "pclass" column. Then, we take the values in the "fare" column and pair them with the corresponding values in the "pclass" column for each row, which are "1", "2", or "3". The values in the "fare" column that correspond to the same values in the "pclass" column will be evaluated using .describe(). The .describe() function will calculate the count, mean, std, min, 25%, 50% (median), 75%, and max values of the "fare" column entries for each unique value in the "pclass" column.

### 2. Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?

### Answer:
It is fundamentally different because when we use df.describe(), the count values are different because there are missing values in the DataFrame and the number of missing values in each column is different, which makes the count value for every column different. Meanwhile, when we use df.groupby("col1")["col2"].describe(), the count values are different because the total number of the same value in the "col1" column is different for each unique value in the "col1" column.

### 3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT

### Answer:
It's easier for me to fix errors with ChatGPT rather than Google Search because with ChatGPT, I can directly get the answer, while with Google Search I have to search for relevant articles and then search for the answer within those long articles.

### ChatGPT's link for question 8 : https://chatgpt.com/share/66e354f6-2af8-800a-a475-34bccb09ebd0
### The summary is in the ChatGPT

# 9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

### Answer: 
Yes